invoices/kv_sql_migration_test.go

```diff
@@ -0,0 +1,124 @@
+package invoices_test
+
+import (
+	"context"
+	"database/sql"
+	"testing"
+	"time"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/clock"
+	invpkg "github.com/lightningnetwork/lnd/invoices"
+	"github.com/lightningnetwork/lnd/sqldb"
+	"github.com/stretchr/testify/require"
+)
+
+// TestMigrationWithChannelDB tests the migration of invoices from a bolt backed
+// channel.db to a SQL database. Note that this test does not attempt to be a
+// complete migration test for all invoice types but rather is added as a tool
+// for developers and users to debug invoice migration issues with an actual
+// channel.db file.
+func TestMigrationWithChannelDB(t *testing.T) {
+	// First create a shared Postgres instance so we don't spawn a new
+	// docker container for each test.
+	pgFixture := sqldb.NewTestPgFixture(
+		t, sqldb.DefaultPostgresFixtureLifetime,
+	)
+	t.Cleanup(func() {
+		pgFixture.TearDown(t)
+	})
+
+	makeSQLDB := func(t *testing.T, sqlite bool) *invpkg.SQLStore {
+		var db *sqldb.BaseDB
+		if sqlite {
+			db = sqldb.NewTestSqliteDB(t).BaseDB
+		} else {
+			db = sqldb.NewTestPostgresDB(t, pgFixture).BaseDB
+		}
+
+		executor := sqldb.NewTransactionExecutor(
+			db, func(tx *sql.Tx) invpkg.SQLInvoiceQueries {
+				return db.WithTx(tx)
+			},
+		)
+
+		testClock := clock.NewTestClock(time.Unix(1, 0))
+
+		return invpkg.NewSQLStore(executor, testClock)
+	}
+
+	migrationTest := func(t *testing.T, store1 *channeldb.DB, sqlite bool) {
+		store2 := makeSQLDB(t, sqlite)
+		ctxb := context.Background()
+
+		const batchSize = 11
+		err := invpkg.MigrateInvoicesToSQL(
+			ctxb, store1.Backend, store1, store2, batchSize,
+		)
+		require.NoError(t, err)
+
+		// MigrateInvoices will check if the inserted invoice equals to
+		// the migrated one, but as a sanity check, we'll also fetch the
+		// invoices from the store and compare them to the original
+		// invoices.
+		query := invpkg.InvoiceQuery{
+			IndexOffset: 0,
+			// As a sanity check, fetch more invoices than we have
+			// to ensure that we did not add any extra invoices.
+			NumMaxInvoices: 9999,
```

Comment:
do we know how much the db had initially would be an important information here.

Reply:
We don't really have a function to tell the number of invoices and I think it'd a waste to add it just for this test. We could query all and count but if someone were to test with a large channeldb they might run out of memory. I think it's fine to keep this constant and if needed can be changed at test time. wdyt?

Reply:
Done ptal.